### PR TITLE
Chroma search optimization

### PR DESCRIPTION
**Pull request by NaderMahdi**
## Description
Perform the Chroma search algorithm just before the last md stage and only when at least one intra mode make it to this stage.
This change is enabled only in encode mode M1 for the moment.

## Related issue
#1047 

## Performance
* Bdrate 0.02% Bdrate loss
* Speed 3.5% speed up
* Memory 1% increase

### CODE REVIEW

File: Source/Lib/Encoder/Codec/EbEncInterPrediction.c [hunk 1]
```diff
@@ -175,82 +172,82 @@ uint64_t av1_wedge_sse_from_residuals_c(const int16_t *r1, const int16_t *d, con
...
-        },
-        {
-                0.000000,    0.000000,    0.000000,    0.000000,    0.000000,    0.000000,    0.000000,
-                0.000000,    0.000000,    0.000000,    0.000000,    8.998436,    9.439592,    9.731837,
-                10.865931,   11.561347,   12.578139,   14.205101,   16.770584,   19.094853,   21.330863,
-                23.298907,   26.901921,   34.501017,   57.891733,   112.234763,  194.853189,  288.302032,
-                380.499422,  472.625309,  560.226809,  647.928463,  734.155122,  817.489721,  906.265783,
-                999.260562,  1094.489206, 1197.062998, 1293.296825, 1378.926484, 1472.760990, 1552.663779,
-                1635.196884, 1692.451951, 1759.741063, 1822.162720, 1916.515921, 1966.686071, 2031.647506,
-                2033.700134, 2087.847688, 2161.688858, 2242.536028, 2334.023491, 2436.337802, 2549.665519,
```
Comment (hassount) [inline Source/Lib/Encoder/Codec/EbEncInterPrediction.c]: This file contains a lot of formatting changes and only a few lines of code that were changed. can we revert back the formatting changes and leave only the additional code ?

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 2]
```diff
@@ -2746,6 +2759,9 @@ static const uint8_t intra_area_th_class_1[MAX_HIERARCHICAL_LEVEL][MAX_TEMPORAL_
 #define CHROMA_MODE_1  1 // Fast chroma search @ MD
 #define CHROMA_MODE_2  2 // Chroma blind @ MD + CFL @ EP
 #define CHROMA_MODE_3  3 // Chroma blind @ MD + no CFL @ EP
+#if MOVE_OPT
+#define CHROMA_MODE_01  4 // Full chroma search @ last MD stage.
+#endif
```
Comment (hassount) [inline Source/Lib/Common/Codec/EbDefinitions.h]: best to fix this now, we need to shift the CHROMA_MODE accordingly

Comment (NaderMahdi) [reply Source/Lib/Common/Codec/EbDefinitions.h]: I didn't get you point. This is a modified chroma0 mode. it is better to keep it this way.

File: Source/Lib/Encoder/Codec/EbEncInterPrediction.c (hunk 1 above)
Comment (NaderMahdi) [reply Source/Lib/Encoder/Codec/EbEncInterPrediction.c]: It is not easy to revert back  and re-base, I guess.

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 2 above)
Comment (hassount) [reply Source/Lib/Common/Codec/EbDefinitions.h]: if it's a modified chroma mode 0 then we call it CHROMA_MODE_1 and shift the other options by 1. such would make a more readable code

File: Source/Lib/Encoder/Codec/EbEncInterPrediction.c (hunk 1 above)
Comment (hassount) [reply Source/Lib/Encoder/Codec/EbEncInterPrediction.c]: if you put the two files side by side with a diff compare tool with the base master, you can use the master file when applicable and the branch parts when applicable then commit those changes that would give you a soft manual revert. No rebasing yet, but you might have to do it ultimately as there are conflicts with the master on this.

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 3]
```diff
@@ -34,6 +34,12 @@
 extern "C" {
 #endif
 
+#define CHROMA_SEARCH_OPT        1 // Move chroma search to be done on the best intra candidate survived from MD stage 2
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbDefinitions.h]: How about: 
Chroma search to be performed for the best surviving intra candidate from MD stage 2

File: Source/Lib/Common/Codec/EbDefinitions.h [hunk 4]
```diff
@@ -34,6 +34,12 @@
...
+#if CHROMA_SEARCH_OPT
+#define INFR_OPT                 1 // lossless
+#define MOVE_OPT                 1 // semi-lossless
+#define COMP_OPT                 1 // lossy
```
Comment (fouedb) [inline Source/Lib/Common/Codec/EbDefinitions.h]: Explain the difference between semi-lossless and lossy.

File: Source/Lib/Common/Codec/EbDefinitions.h (hunk 2 above)
Comment (mafonsoN) [reply Source/Lib/Common/Codec/EbDefinitions.h]: Having both CHROMA_MODE_1 and CHROMA_MODE_01 is very confusing.

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 5]
```diff
@@ -5054,7 +5105,28 @@ void md_stage_3(PictureControlSet *pcs_ptr, SuperBlock *sb_ptr, BlkStruct *blk_p
                 return;
             }
         }
+#if MOVE_OPT
+        if (context_ptr->chroma_level == CHROMA_MODE_01) {
+            if (context_ptr->blk_geom->sq_size < 128) {
```
Comment (mafonsoN) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: Perhaps you can use MAX_SB_SIZE instead

File: Source/Lib/Encoder/Codec/EbProductCodingLoop.c [hunk 6]
```diff
@@ -5565,6 +5637,24 @@ void order_nsq_table(PictureControlSet *pcs_ptr, ModeDecisionContext *context_pt
...
 }
+#if MOVE_OPT
+void init_chroma_mode(ModeDecisionContext   *context_ptr) {
+    context_ptr->uv_search_path = EB_TRUE;
+    EbBool use_angle_delta = av1_use_angle_delta(context_ptr->blk_geom->bsize);
+    for (uint8_t intra_mode = DC_PRED; intra_mode <= PAETH_PRED; ++intra_mode) {
+        uint8_t angleDeltaCandidateCount = (use_angle_delta && av1_is_directional_mode((PredictionMode)intra_mode)) ? 7 : 1;
+        uint8_t angle_delta_shift = 1;
+        for (uint8_t angleDeltaCounter = 0; angleDeltaCounter < angleDeltaCandidateCount; ++angleDeltaCounter) {
+            int32_t angle_delta = CLIP(angle_delta_shift * (angleDeltaCandidateCount == 1 ? 0 : angleDeltaCounter - (angleDeltaCandidateCount >> 1)), -MAX_ANGLE_DELTA, MAX_ANGLE_DELTA);
```
Comment (mafonsoN) [inline Source/Lib/Encoder/Codec/EbProductCodingLoop.c]: This line is too long

**Comment by hassount** [comment]
#1071

**Pull request closed without merging**